### PR TITLE
[HOLD] feat: anonymous user claim account

### PR DESCRIPTION
**Pull request by thiagobrez**
### Fixed Issues

$ https://github.com/Expensify/App/issues/18321
PROPOSAL: GH_LINK_ISSUE(COMMENT)

### Tests

- [ ] Verify that no errors appear in the JS console
- [ ] Sign out
- [ ] Open a public room (e.g. https://staging.new.expensify.com/r/6469015542180863, remember to switch the URL to localhost)
- [ ] Verify that you're anonymous (cant do any actions, options to sign in instead)
- [ ] When clicking to sign in (or trying to do a chat action) a modal to sign in should open
- [ ] Sign-in normally
- [ ] Modal should disappear, session and chat should refresh, everything should work normally

### Offline tests

If offline, Modal should appear but user won't be able to sign in. Sign In Page already handles offline behavior.

### QA Steps

- [ ] Verify that no errors appear in the JS console
- [ ] Sign out
- [ ] Open a public room (e.g. https://staging.new.expensify.com/r/6469015542180863, remember to switch the URL to localhost)
- [ ] Verify that you're anonymous (cant do any actions, options to sign in instead)
- [ ] When clicking to sign in (or trying to do a chat action) a modal to sign in should open
- [ ] Sign-in normally
- [ ] Modal should disappear, session and chat should refresh, everything should work normally

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [X] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>
<video src="https://github.com/Expensify/App/assets/26878038/6f3a9961-9b4d-4b7c-a676-261dd5930ae2" />
</details>

<details>
<summary>Mobile Web - Chrome</summary>
<video src="https://github.com/Expensify/App/assets/26878038/478f9764-0ebe-4777-a634-b436d4e80842" />
</details>

<details>
<summary>Mobile Web - Safari</summary>
<video src="https://github.com/Expensify/App/assets/26878038/75f237b4-0587-4fe8-9d4c-67f26f64e5fc" />
</details>

<details>
<summary>Desktop</summary>

N/A

</details>

<details>
<summary>iOS</summary>

N/A

</details>

<details>
<summary>Android</summary>

N/A

</details>

### CODE REVIEW

File: src/components/AnonymousReportFooter.js [hunk 1]
```diff
@@ -44,7 +44,7 @@ const AnonymousReportFooter = (props) => (
                     medium
                     success
                     text={props.translate('common.signIn')}
-                    onPress={() => Session.signOutAndRedirectToSignIn()}
+                    onPress={() => SignInModalActions.showSignInModal()}
```
Comment (gedu) [inline src/components/AnonymousReportFooter.js:47]: you can just do 

```
      onPress={SignInModalActions.showSignInModal}
```

File: src/pages/home/sidebar/SidebarLinks.js [hunk 2]
```diff
@@ -186,7 +187,7 @@ class SidebarLinks extends React.Component {
                                     medium
                                     success
                                     text={this.props.translate('common.signIn')}
-                                    onPress={() => Session.signOutAndRedirectToSignIn()}
+                                    onPress={() => SignInModalActions.showSignInModal()}
```
Comment (gedu) [inline src/pages/home/sidebar/SidebarLinks.js:190]: dito 
```
onPress={SignInModalActions.showSignInModal}
```

**Comment by thiagobrez** [comment]
Waiting on `ClaimAnonymousAccount` backend command to be ready, per [this comment](https://github.com/Expensify/App/issues/18321#issuecomment-1577125160)

**Comment by marcochavezf** [comment]
Thanks @thiagobrez! I will apply the `Design` label to get some 👀 on the design of the modal.

**Comment by marcochavezf** [comment]
Oops, seems it only works on the issue and not in the PRs

**Comment by shawnborton** [comment]
@marcochavezf I see you asked for my feedback on this one

Isn't this missing the bottom banner from this screenshot here? 
image

**Comment by marcochavezf** [comment]
Oh that's true, I realized that we're not showing the banner when the chat is too long, I will check it out. And what about the modal when the anonymous user clicks on the "Sign in" button? 

Screenshot 2023-06-08 at 11 49 07

**Comment by shawnborton** [comment]
Oh wow, I didn't even realize that was a modal. I gotta say, I am not a fan of that at all. Why not just redirect them to the sign in page and redirect them back once they sign in?

**Comment by shawnborton** [comment]
Was the modal approach discussed somewhere in Slack or anything?

**Comment by marcochavezf** [comment]
> Was the modal approach discussed somewhere in Slack or anything?

Ah sorry if that wasn't discussed widely in Slack, but I added it in the design doc [here](https://docs.google.com/document/d/1kFrc-urE13h3-XbyRmu5jIqyr9CuT02SqO4tnzsSuAY/edit#bookmark=id.i7wvgnqbddgn). I added thinking that this would have a better UX flow (a less interrupted flow as also @thiagobrez mentioned [here](https://github.com/Expensify/App/issues/18321#issuecomment-1568610398)) and because technically it's a bit challenging to redirect to the sign-in page and then let the user navigate back to the public room (we need to force the anonymous user to log out to show the sign-in page again).

Alternatively, we could show the sign-in page as a modal but in full-screen mode (giving the impression that it's not a modal and the user is redirected to the sign-in page). Thoughts?

**Comment by shawnborton** [comment]
I think the modal experience feels a bit strange personally. I would either make it a right-hand-nav modal where the modal is slim, or just redirect the user to sign in and then redirect them back once signed in. Either way, I think it's worth bringing this one back to Slack for a wider discussion. 

That being said, my vote is just to do a full page redirect and avoid the modal weirdness.

**Comment by marcochavezf** [comment]
Putting this PR on hold until we have the sign-in page in the RHP and the claim anonymous account command (more details [here](https://github.com/Expensify/App/issues/18321#issuecomment-1604619362))

**Pull request closed without merging**